src/expr.rs

```diff
@@ -1102,10 +1102,31 @@ impl<'a> Rewrite for ControlFlow<'a> {
         };
         let block_str = {
             let old_val = context.is_if_else_block.replace(self.else_block.is_some());
-            let result =
-                rewrite_block_with_visitor(context, "", self.block, None, None, block_shape, true);
+            let max_width = if context.config.single_line_simple_if() {
+                std::cmp::min(
+                    shape.width,
+                    context.config.single_line_simple_if_max_width(),
+                )
+            } else {
+                shape.width
+            };
+            let available_space = max_width.saturating_sub(used_width);
+            let allow_single_line = allow_single_line_if(&cond_str, self.block)
+                && self.keyword == "if"
+                && available_space > 0;
+
+            let mut result = if allow_single_line && context.config.single_line_simple_if() {
+                rewrite_block_inner(self.block, None, None, true, context, block_shape)?
+            } else {
+                rewrite_block_with_visitor(context, "", self.block, None, None, block_shape, true)?
+            };
+
+            let block_exceeds_width = result.len() > available_space;
+            if allow_single_line && !result.contains('\n') && block_exceeds_width {
+                result = rewrite_block_inner(self.block, None, None, false, context, block_shape)?;
+            }
             context.is_if_else_block.replace(old_val);
-            result?
+            result
```

Comment:
The `ControlFlow` type abstracts over a lot of different kinds of control flow rewriting. It probably makes sense to move the `if` specific logic out into into a standalone function.


```rust
let result = if self.keyword == "if" {
    call_if_formatting_function()
} else {
    rewrite_block_with_visitor(context, "", self.block, None, None, block_shape, true);
};
```